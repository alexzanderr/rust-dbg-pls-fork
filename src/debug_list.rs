use crate::{DebugPls, Formatter};

/// A helper designed to assist with creation of
/// [`DebugPls`] implementations for list-like structures.
///
/// # Examples
///
/// ```rust
/// use dbg_pls::{pretty, DebugPls, Formatter};
///
/// struct Foo(Vec<i32>);
///
/// impl DebugPls for Foo {
///     fn fmt(&self, f: Formatter<'_>) {
///         f.debug_list().entries(&self.0).finish()
///     }
/// }
///
/// let value = Foo(vec![10, 11]);
/// assert_eq!(format!("{}", pretty(&value)), "[10, 11]");
/// ```
pub struct DebugList<'a> {
    formatter: Formatter<'a>,
    expr: syn::ExprArray,
}

impl<'a> DebugList<'a> {
    pub(crate) fn new(formatter: Formatter<'a>) -> Self {
        DebugList {
            formatter,
            expr: syn::ExprArray {
                attrs: vec![],
                bracket_token: syn::token::Bracket::default(),
                elems: syn::punctuated::Punctuated::default(),
            },
        }
    }

    /// Adds a new entry to the list output.
    #[must_use]
    pub fn entry(mut self, entry: &dyn DebugPls) -> Self {
        self.expr.elems.push(Formatter::process(entry));
        self
    }

    /// Adds all the entries to the list output.
    #[must_use]
    pub fn entries<D, I>(mut self, entries: I) -> Self
    where
        D: DebugPls,
        I: IntoIterator<Item = D>,
    {
        self.extend(entries);
        self
    }

    /// Closes off the list
    pub fn finish(self) {
        self.formatter.write_expr(self.expr);
    }
}

impl<'f, D: DebugPls> Extend<D> for DebugList<'f> {
    fn extend<T: IntoIterator<Item = D>>(&mut self, iter: T) {
        self.expr
            .elems
            .extend(iter.into_iter().map(|entry| Formatter::process(&entry)));
    }
}
