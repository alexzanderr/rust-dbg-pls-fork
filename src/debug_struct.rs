use syn::__private::Span;

use crate::{DebugPls, Formatter};

pub struct DebugStruct<'a> {
    formatter: Formatter<'a>,
    expr: syn::ExprStruct,
}

impl<'a> DebugStruct<'a> {
    pub(crate) fn new(formatter: Formatter<'a>, name: &str) -> Self {
        DebugStruct {
            formatter,
            expr: syn::ExprStruct {
                attrs: vec![],
                path: syn::Ident::new(name, Span::call_site()).into(),
                brace_token: syn::token::Brace::default(),
                fields: syn::punctuated::Punctuated::new(),
                dot2_token: None,
                rest: None,
            },
        }
    }

    pub fn field(mut self, name: &str, value: &dyn DebugPls) -> DebugStruct<'a> {
        self.expr.fields.push(syn::FieldValue {
            expr: Formatter::process(value),
            attrs: vec![],
            member: syn::Member::Named(syn::Ident::new(name, Span::call_site())),
            colon_token: Some(syn::token::Colon::default()),
        });
        self
    }

    pub fn finish(self) {
        *self.formatter.expr = syn::Expr::Struct(self.expr);
    }

    pub fn finish_non_exhaustive(mut self) {
        self.expr.dot2_token = Some(syn::token::Dot2::default());
        self.finish()
    }
}
