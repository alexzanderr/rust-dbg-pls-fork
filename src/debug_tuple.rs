use crate::{DebugPls, Formatter};

pub struct DebugTuple<'a> {
    formatter: Formatter<'a>,
    expr: syn::ExprTuple,
}

impl<'a> DebugTuple<'a> {
    pub(crate) fn new(formatter: Formatter<'a>) -> Self {
        DebugTuple {
            formatter,
            expr: syn::ExprTuple {
                attrs: vec![],
                paren_token: syn::token::Paren::default(),
                elems: syn::punctuated::Punctuated::new(),
            },
        }
    }

    pub fn field(mut self, value: &dyn DebugPls) -> Self {
        self.expr.elems.push(Formatter::process(value));
        self
    }

    pub fn finish(self) {
        self.formatter.write_expr(self.expr);
    }
}
