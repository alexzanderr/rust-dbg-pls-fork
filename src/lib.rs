//! Syntax aware debug printing.
//!
//! Makes use of `syn` and `prettyplease` in order to provide the most
//! canonincal rust debug lines as possible, quickly.
//!
//! # Example usage
//!
//! ```
//! use dbg_pls::{debug, DebugPls};
//!
//! #[derive(DebugPls, Copy, Clone)]
//! pub struct Demo {
//!     foo: i32,
//!     bar: &'static str,
//! }
//!
//! let mut val = [Demo { foo: 5, bar: "hello" }; 10];
//! val[6].bar = "Hello, world! I am a very long string";
//!
//! let output = format!("{}", debug(&val));
//! let expected = r#"[
//!     Demo { foo: 5, bar: "hello" },
//!     Demo { foo: 5, bar: "hello" },
//!     Demo { foo: 5, bar: "hello" },
//!     Demo { foo: 5, bar: "hello" },
//!     Demo { foo: 5, bar: "hello" },
//!     Demo { foo: 5, bar: "hello" },
//!     Demo {
//!         foo: 5,
//!         bar: "Hello, world! I am a very long string",
//!     },
//!     Demo { foo: 5, bar: "hello" },
//!     Demo { foo: 5, bar: "hello" },
//!     Demo { foo: 5, bar: "hello" },
//! ]"#;
//!
//! assert_eq!(output, expected);
//! ```
//!
//! # Example with highlighting
//!
//! ```
//! use dbg_pls::{color, DebugPls};
//!
//! #[derive(DebugPls, Copy, Clone)]
//! pub struct Demo {
//!     foo: i32,
//!     bar: &'static str,
//! }
//!
//! let mut val = [Demo { foo: 5, bar: "hello" }; 10];
//! val[6].bar = "Hello, world! I am a very long string";
//!
//! println!("{}", color(&val));
//! ```
//! Outputs:
//!
//! ![](https://raw.githubusercontent.com/conradludgate/dbg-pls/5dee03187a3f83693739e0288d56da5980e1d486/readme/highlighted.png)

use syn::__private::{Span, TokenStream2};

mod impls;

mod debug_list;
mod debug_struct;
mod debug_tuple;
use debug_list::DebugList;
use debug_struct::DebugStruct;
use debug_tuple::DebugTuple;

#[cfg(feature = "pretty")]
mod pretty;
#[cfg(feature = "pretty")]
pub use pretty::debug;

#[cfg(feature = "colors")]
mod colors;
#[cfg(feature = "colors")]
pub use colors::color;

pub use dbg_pls_derive::DebugPls;

/// Syntax aware pretty-printed debug formatting.
///
/// `DebugPls` should format the output in a programmer-facing, debugging context.
///
/// Generally speaking, you should just `derive` a `Debug` implementation.
pub trait DebugPls {
    fn fmt(&self, f: Formatter<'_>);
}

pub struct Formatter<'a> {
    expr: &'a mut syn::Expr,
}

impl<'a> Formatter<'a> {
    pub(crate) fn process(value: &dyn DebugPls) -> syn::Expr {
        let mut expr = syn::Expr::Verbatim(TokenStream2::new());
        value.fmt(Formatter { expr: &mut expr });
        expr
    }

    pub fn write_expr(self, expr: impl Into<syn::Expr>) {
        *self.expr = expr.into();
    }

    pub fn debug_struct(self, name: &str) -> DebugStruct<'a> {
        DebugStruct::new(self, name)
    }
    pub fn debug_tuple(self, name: &str) -> DebugTuple<'a> {
        DebugTuple::new(self, name)
    }
    pub fn debug_list(self) -> DebugList<'a> {
        DebugList::new(self)
    }
    pub fn debug_ident(self, name: &str) {
        let path: syn::Path = syn::Ident::new(name, Span::call_site()).into();
        self.write_expr(syn::ExprPath {
            attrs: vec![],
            qself: None,
            path,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(DebugPls, Copy, Clone)]
    #[dbg_pls(crate = "crate")]
    pub struct Demo {
        foo: i32,
        bar: &'static str,
    }

    #[test]
    fn debug_struct() {
        let val = Demo {
            foo: 5,
            bar: "hello",
        };
        assert_eq!(debug(&val).to_string(), r#"Demo { foo: 5, bar: "hello" }"#);
    }

    #[test]
    fn debug_struct_big() {
        let val = Demo {
            foo: 5,
            bar: "Hello, world! I am a very long string",
        };
        assert_eq!(
            debug(&val).to_string(),
            r#"Demo {
    foo: 5,
    bar: "Hello, world! I am a very long string",
}"#
        );
    }

    #[test]
    fn debug_nested_struct() {
        let mut val = [Demo {
            foo: 5,
            bar: "hello",
        }; 10];
        val[6].bar = "Hello, world! I am a very long string";

        assert_eq!(
            debug(&val).to_string(),
            r#"[
    Demo { foo: 5, bar: "hello" },
    Demo { foo: 5, bar: "hello" },
    Demo { foo: 5, bar: "hello" },
    Demo { foo: 5, bar: "hello" },
    Demo { foo: 5, bar: "hello" },
    Demo { foo: 5, bar: "hello" },
    Demo {
        foo: 5,
        bar: "Hello, world! I am a very long string",
    },
    Demo { foo: 5, bar: "hello" },
    Demo { foo: 5, bar: "hello" },
    Demo { foo: 5, bar: "hello" },
]"#
        );
    }
}
