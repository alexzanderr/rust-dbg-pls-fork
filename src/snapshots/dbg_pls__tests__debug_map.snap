---
source: src/lib.rs
expression: "BTreeMap::from([(\"hello\", 60), (\"Hello, world! I am a very long string\", 12)])"
---
{
    ["Hello, world! I am a very long string"] = 12;
    ["hello"] = 60;
}
